File: open_source_projects/Deepbox-main/benchmarks/deepbox/01-dataframe.ts
/**
 * Benchmark 01 — DataFrame Operations
 * Deepbox vs Pandas
 */

import { DataFrame } from "deepbox/dataframe";
import { createSuite, footer, header, run } from "../utils";

const suite = createSuite("dataframe");
header("Benchmark 01 — DataFrame Operations");

// ── Data generators ──────────────────────────────────────

function seededRng(seed: number) {
	let s = seed >>> 0;
	return () => {
		s = (s * 1664525 + 1013904223) >>> 0;
		return s / 2 ** 32;
	};
}

function makeNumericDF(nRows: number, nCols: number, seed: number): DataFrame {
	const rand = seededRng(seed);
	const data: Record<string, number[]> = {};
	for (let c = 0; c < nCols; c++) {
		const col: number[] = [];
		for (let r = 0; r < nRows; r++) col.push(rand() * 100);
		data[`col${c}`] = col;
	}
	return new DataFrame(data);
}

function makeMixedDF(nRows: number, seed: number): DataFrame {
	const rand = seededRng(seed);
	const names: string[] = [];
	const ages: number[] = [];
	const scores: number[] = [];
	const categories: string[] = [];
	const values: number[] = [];
	const cats = ["A", "B", "C", "D", "E"];
	for (let i = 0; i < nRows; i++) {
		names.push(`person_${i}`);
		ages.push(Math.floor(rand() * 60) + 18);
		scores.push(rand() * 100);
		categories.push(cats[Math.floor(rand() * cats.length)]);
		values.push(rand() * 1000);
	}
	return new DataFrame({
		name: names,
		age: ages,
		score: scores,
		category: categories,
		value: values,
	});
}

const _df100 = makeMixedDF(100, 42);
const df1k = makeMixedDF(1000, 42);
const df10k = makeMixedDF(10000, 42);
const df50k = makeMixedDF(50000, 42);
const numDf100 = makeNumericDF(100, 5, 42);
const numDf1k = makeNumericDF(1000, 5, 42);
const numDf10k = makeNumericDF(10000, 5, 42);

// ── Creation ─────────────────────────────────────────────

run(suite, "create", "100 rows", () => makeMixedDF(100, 99));
run(suite, "create", "1K rows", () => makeMixedDF(1000, 99));
run(suite, "create", "10K rows", () => makeMixedDF(10000, 99));
run(suite, "create (numeric)", "100x5", () => makeNumericDF(100, 5, 99));
run(suite, "create (numeric)", "1Kx5", () => makeNumericDF(1000, 5, 99));
run(suite, "create (numeric)", "10Kx5", () => makeNumericDF(10000, 5, 99));

// ── Select ───────────────────────────────────────────────

run(suite, "select (1 col)", "1K rows", () => df1k.select(["score"]));
run(suite, "select (1 col)", "10K rows", () => df10k.select(["score"]));
run(suite, "select (3 cols)", "1K rows", () => df1k.select(["name", "age", "score"]));
run(suite, "select (3 cols)", "10K rows", () => df10k.select(["name", "age", "score"]));

// ── Filter ───────────────────────────────────────────────

run(suite, "filter (numeric >)", "1K rows", () => df1k.filter((r) => (r.age as number) > 40));
run(suite, "filter (numeric >)", "10K rows", () => df10k.filter((r) => (r.age as number) > 40));
run(suite, "filter (string ==)", "1K rows", () => df1k.filter((r) => r.category === "A"));
run(suite, "filter (string ==)", "10K rows", () => df10k.filter((r) => r.category === "A"));
run(suite, "filter (compound)", "1K rows", () =>
	df1k.filter((r) => (r.age as number) > 30 && (r.score as number) > 50)
);
run(suite, "filter (compound)", "10K rows", () =>
	df10k.filter((r) => (r.age as number) > 30 && (r.score as number) > 50)
);

// ── Sort ─────────────────────────────────────────────────

run(suite, "sort (single col)", "1K rows", () => df1k.sort("age"));
run(suite, "sort (single col)", "10K rows", () => df10k.sort("age"));
run(suite, "sort (descending)", "1K rows", () => df1k.sort("score", false));
run(suite, "sort (descending)", "10K rows", () => df10k.sort("score", false));

// ── GroupBy ──────────────────────────────────────────────

run(suite, "groupBy + sum", "1K rows", () => df1k.groupBy("category").sum());
run(suite, "groupBy + sum", "10K rows", () => df10k.groupBy("category").sum());
run(suite, "groupBy + mean", "1K rows", () => df1k.groupBy("category").mean());
run(suite, "groupBy + mean", "10K rows", () => df10k.groupBy("category").mean());
run(suite, "groupBy + count", "1K rows", () => df1k.groupBy("category").count());
run(suite, "groupBy + count", "10K rows", () => df10k.groupBy("category").count());
run(suite, "groupBy + min", "1K rows", () => df1k.groupBy("category").min());
run(suite, "groupBy + min", "10K rows", () => df10k.groupBy("category").min());
run(suite, "groupBy + max", "1K rows", () => df1k.groupBy("category").max());
run(suite, "groupBy + max", "10K rows", () => df10k.groupBy("category").max());

// ── Head / Tail ──────────────────────────────────────────

run(suite, "head(10)", "10K rows", () => df10k.head(10));
run(suite, "head(10)", "50K rows", () => df50k.head(10));
run(suite, "tail(10)", "10K rows", () => df10k.tail(10));
run(suite, "tail(10)", "50K rows", () => df50k.tail(10));

// ── Loc / Iloc ───────────────────────────────────────────

run(suite, "iloc(0)", "1K rows", () => df1k.iloc(0));
run(suite, "iloc(500)", "1K rows", () => df1k.iloc(500));
run(suite, "iloc(0)", "10K rows", () => df10k.iloc(0));
run(suite, "iloc(5000)", "10K rows", () => df10k.iloc(5000));

// ── Join ─────────────────────────────────────────────────

const leftDf = new DataFrame({
	key: Array.from({ length: 500 }, (_, i) => i % 50),
	val_l: Array.from({ length: 500 }, (_, i) => i),
});
const rightDf = new DataFrame({
	key: Array.from({ length: 50 }, (_, i) => i),
	val_r: Array.from({ length: 50 }, (_, i) => i * 10),
});
const leftDf2k = new DataFrame({
	key: Array.from({ length: 2000 }, (_, i) => i % 100),
	val_l: Array.from({ length: 2000 }, (_, i) => i),
});
const rightDf100 = new DataFrame({
	key: Array.from({ length: 100 }, (_, i) => i),
	val_r: Array.from({ length: 100 }, (_, i) => i * 10),
});

run(suite, "join (inner)", "500×50", () => leftDf.join(rightDf, "key", "inner"));
run(suite, "join (left)", "500×50", () => leftDf.join(rightDf, "key", "left"));
run(suite, "join (inner)", "2K×100", () => leftDf2k.join(rightDf100, "key", "inner"));
run(suite, "join (left)", "2K×100", () => leftDf2k.join(rightDf100, "key", "left"));

// ── Concat ───────────────────────────────────────────────

run(suite, "concat (axis=0)", "2×1K", () => df1k.concat(df1k, 0));
run(suite, "concat (axis=0)", "2×10K", () => df10k.concat(df10k, 0));

// ── FillNa / DropNa ─────────────────────────────────────

const dfWithNa1k = new DataFrame({
	a: Array.from({ length: 1000 }, (_, i) => (i % 5 === 0 ? null : i)),
	b: Array.from({ length: 1000 }, (_, i) => (i % 3 === 0 ? null : i * 2)),
});
const dfWithNa10k = new DataFrame({
	a: Array.from({ length: 10000 }, (_, i) => (i % 5 === 0 ? null : i)),
	b: Array.from({ length: 10000 }, (_, i) => (i % 3 === 0 ? null : i * 2)),
});

run(suite, "fillna(0)", "1K rows", () => dfWithNa1k.fillna(0));
run(suite, "fillna(0)", "10K rows", () => dfWithNa10k.fillna(0));
run(suite, "dropna", "1K rows", () => dfWithNa1k.dropna());
run(suite, "dropna", "10K rows", () => dfWithNa10k.dropna());

// ── Describe ─────────────────────────────────────────────

run(suite, "describe", "100x5", () => numDf100.describe());
run(suite, "describe", "1Kx5", () => numDf1k.describe());
run(suite, "describe", "10Kx5", () => numDf10k.describe());

// ── Correlation ──────────────────────────────────────────

run(suite, "corr", "100x5", () => numDf100.corr());
run(suite, "corr", "1Kx5", () => numDf1k.corr());

// ── Drop ─────────────────────────────────────────────────

run(suite, "drop (1 col)", "1K rows", () => df1k.drop(["value"]));
run(suite, "drop (2 cols)", "10K rows", () => df10k.drop(["value", "name"]));

footer(suite, "deepbox-dataframe.json");

File: open_source_projects/Deepbox-main/benchmarks/python/01_dataframe.py
"""
Benchmark 01 — DataFrame Operations
Pandas
"""

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

import numpy as np
import pandas as pd
from utils import run, create_suite, header, footer

suite = create_suite("dataframe", "Pandas")
header("Benchmark 01 — DataFrame Operations", "Pandas")

# ── Data generators ──────────────────────────────────────

def make_mixed_df(n, seed):
    rng = np.random.RandomState(seed)
    cats = ["A", "B", "C", "D", "E"]
    return pd.DataFrame({
        "name": [f"person_{i}" for i in range(n)],
        "age": rng.randint(18, 78, n),
        "score": rng.rand(n) * 100,
        "category": [cats[i % 5] for i in rng.randint(0, 5, n)],
        "value": rng.rand(n) * 1000,
    })

def make_numeric_df(n, ncols, seed):
    rng = np.random.RandomState(seed)
    return pd.DataFrame({f"col{c}": rng.rand(n) * 100 for c in range(ncols)})

df100 = make_mixed_df(100, 42)
df1k = make_mixed_df(1000, 42)
df10k = make_mixed_df(10000, 42)
df50k = make_mixed_df(50000, 42)
numDf100 = make_numeric_df(100, 5, 42)
numDf1k = make_numeric_df(1000, 5, 42)
numDf10k = make_numeric_df(10000, 5, 42)

# ── Creation ─────────────────────────────────────────────

run(suite, "create", "100 rows", lambda: make_mixed_df(100, 99))
run(suite, "create", "1K rows", lambda: make_mixed_df(1000, 99))
run(suite, "create", "10K rows", lambda: make_mixed_df(10000, 99))
run(suite, "create (numeric)", "100x5", lambda: make_numeric_df(100, 5, 99))
run(suite, "create (numeric)", "1Kx5", lambda: make_numeric_df(1000, 5, 99))
run(suite, "create (numeric)", "10Kx5", lambda: make_numeric_df(10000, 5, 99))

# ── Select ───────────────────────────────────────────────

run(suite, "select (1 col)", "1K rows", lambda: df1k[["score"]])
run(suite, "select (1 col)", "10K rows", lambda: df10k[["score"]])
run(suite, "select (3 cols)", "1K rows", lambda: df1k[["name", "age", "score"]])
run(suite, "select (3 cols)", "10K rows", lambda: df10k[["name", "age", "score"]])

# ── Filter ───────────────────────────────────────────────

run(suite, "filter (numeric >)", "1K rows", lambda: df1k[df1k["age"] > 40])
run(suite, "filter (numeric >)", "10K rows", lambda: df10k[df10k["age"] > 40])
run(suite, "filter (string ==)", "1K rows", lambda: df1k[df1k["category"] == "A"])
run(suite, "filter (string ==)", "10K rows", lambda: df10k[df10k["category"] == "A"])
run(suite, "filter (compound)", "1K rows", lambda: df1k[(df1k["age"] > 30) & (df1k["score"] > 50)])
run(suite, "filter (compound)", "10K rows", lambda: df10k[(df10k["age"] > 30) & (df10k["score"] > 50)])

# ── Sort ─────────────────────────────────────────────────

run(suite, "sort (single col)", "1K rows", lambda: df1k.sort_values("age"))
run(suite, "sort (single col)", "10K rows", lambda: df10k.sort_values("age"))
run(suite, "sort (descending)", "1K rows", lambda: df1k.sort_values("score", ascending=False))
run(suite, "sort (descending)", "10K rows", lambda: df10k.sort_values("score", ascending=False))

# ── GroupBy ──────────────────────────────────────────────

run(suite, "groupBy + sum", "1K rows", lambda: df1k.groupby("category").sum(numeric_only=True))
run(suite, "groupBy + sum", "10K rows", lambda: df10k.groupby("category").sum(numeric_only=True))
run(suite, "groupBy + mean", "1K rows", lambda: df1k.groupby("category").mean(numeric_only=True))
run(suite, "groupBy + mean", "10K rows", lambda: df10k.groupby("category").mean(numeric_only=True))
run(suite, "groupBy + count", "1K rows", lambda: df1k.groupby("category").count())
run(suite, "groupBy + count", "10K rows", lambda: df10k.groupby("category").count())
run(suite, "groupBy + min", "1K rows", lambda: df1k.groupby("category").min(numeric_only=True))
run(suite, "groupBy + min", "10K rows", lambda: df10k.groupby("category").min(numeric_only=True))
run(suite, "groupBy + max", "1K rows", lambda: df1k.groupby("category").max(numeric_only=True))
run(suite, "groupBy + max", "10K rows", lambda: df10k.groupby("category").max(numeric_only=True))

# ── Head / Tail ──────────────────────────────────────────

run(suite, "head(10)", "10K rows", lambda: df10k.head(10))
run(suite, "head(10)", "50K rows", lambda: df50k.head(10))
run(suite, "tail(10)", "10K rows", lambda: df10k.tail(10))
run(suite, "tail(10)", "50K rows", lambda: df50k.tail(10))

# ── Loc / Iloc ───────────────────────────────────────────

run(suite, "iloc(0)", "1K rows", lambda: df1k.iloc[0])
run(suite, "iloc(500)", "1K rows", lambda: df1k.iloc[500])
run(suite, "iloc(0)", "10K rows", lambda: df10k.iloc[0])
run(suite, "iloc(5000)", "10K rows", lambda: df10k.iloc[5000])

# ── Join ─────────────────────────────────────────────────

left_df = pd.DataFrame({"key": [i % 50 for i in range(500)], "val_l": list(range(500))})
right_df = pd.DataFrame({"key": list(range(50)), "val_r": [i * 10 for i in range(50)]})
left_df2k = pd.DataFrame({"key": [i % 100 for i in range(2000)], "val_l": list(range(2000))})
right_df100 = pd.DataFrame({"key": list(range(100)), "val_r": [i * 10 for i in range(100)]})

run(suite, "join (inner)", "500×50", lambda: left_df.merge(right_df, on="key", how="inner"))
run(suite, "join (left)", "500×50", lambda: left_df.merge(right_df, on="key", how="left"))
run(suite, "join (inner)", "2K×100", lambda: left_df2k.merge(right_df100, on="key", how="inner"))
run(suite, "join (left)", "2K×100", lambda: left_df2k.merge(right_df100, on="key", how="left"))

# ── Concat ───────────────────────────────────────────────

run(suite, "concat (axis=0)", "2×1K", lambda: pd.concat([df1k, df1k], axis=0, ignore_index=True))
run(suite, "concat (axis=0)", "2×10K", lambda: pd.concat([df10k, df10k], axis=0, ignore_index=True))

# ── FillNa / DropNa ─────────────────────────────────────

dfna1k = pd.DataFrame({
    "a": [None if i % 5 == 0 else i for i in range(1000)],
    "b": [None if i % 3 == 0 else i * 2 for i in range(1000)],
})
dfna10k = pd.DataFrame({
    "a": [None if i % 5 == 0 else i for i in range(10000)],
    "b": [None if i % 3 == 0 else i * 2 for i in range(10000)],
})

run(suite, "fillna(0)", "1K rows", lambda: dfna1k.fillna(0))
run(suite, "fillna(0)", "10K rows", lambda: dfna10k.fillna(0))
run(suite, "dropna", "1K rows", lambda: dfna1k.dropna())
run(suite, "dropna", "10K rows", lambda: dfna10k.dropna())

# ── Describe ─────────────────────────────────────────────

run(suite, "describe", "100x5", lambda: numDf100.describe())
run(suite, "describe", "1Kx5", lambda: numDf1k.describe())
run(suite, "describe", "10Kx5", lambda: numDf10k.describe())

# ── Correlation ──────────────────────────────────────────

run(suite, "corr", "100x5", lambda: numDf100.corr())
run(suite, "corr", "1Kx5", lambda: numDf1k.corr())

# ── Drop ─────────────────────────────────────────────────

run(suite, "drop (1 col)", "1K rows", lambda: df1k.drop(columns=["value"]))
run(suite, "drop (2 cols)", "10K rows", lambda: df10k.drop(columns=["value", "name"]))

footer(suite, "pandas-dataframe.json")
